packages/runtime/test-runtime-utils/src/mocks.ts

```diff
@@ -385,9 +389,92 @@ export class MockContainerRuntimeFactory {
 	}
 
 	private lastProcessedMessage?: ISequencedDocumentMessage;
-	private processFirstMessage() {
-		assert(this.messages.length > 0, "The message queue should not be empty");
 
+	/**
+	 * Catches all clients behind frontIndex up to frontIndex.
+	 */
+	private catchUpClients() {
+		for (const [client, index] of this.clientToIndex) {
+			let mutableIndex = index;
+			while (mutableIndex < this.frontIndex) {
+				const messageAtIndex = JSON.parse(
+					JSON.stringify(this.messages[mutableIndex]),
+				) as ISequencedDocumentMessage;
+				this.minSeq.set(messageAtIndex.clientId, messageAtIndex.referenceSequenceNumber);
+				messageAtIndex.sequenceNumber =
+					this.sequenceNumber - (this.frontIndex - mutableIndex);
+				messageAtIndex.minimumSequenceNumber = this.getMinSeq();
+
+				this.runtimes.get(client)?.process(messageAtIndex);
+
+				mutableIndex++;
+			}
+		}
+
+		this.messages = this.messages.slice(this.frontIndex);
+		this.frontIndex = 0;
+	}
+
+	private catchUpClient(client: string) {
+		let index = this.clientToIndex.get(client);
+		if (index !== undefined) {
+			while (index < this.frontIndex) {
+				const messageAtIndex = JSON.parse(
+					JSON.stringify(this.messages[index]),
+				) as ISequencedDocumentMessage;
+				this.minSeq.set(messageAtIndex.clientId, messageAtIndex.referenceSequenceNumber);
+				messageAtIndex.sequenceNumber = this.sequenceNumber - index;
+				messageAtIndex.minimumSequenceNumber = this.getMinSeq();
+
+				this.runtimes[client].process(messageAtIndex);
+
+				index++;
+			}
+
+			this.clientToIndex.delete(client);
+		}
+	}
+
+	private advanceClient(client: string) {
+		assert(
+			this.clientToIndex[client] === undefined,
+			"Client should be caught up before trying to advance",
+		);
+
+		for (const [clientId, _] of this.runtimes) {
+			if (clientId !== client && this.clientToIndex.get(clientId) === undefined) {
+				this.clientToIndex.set(clientId, this.frontIndex);
+			}
+		}
+
+		const message = JSON.parse(
+			JSON.stringify(this.messages[this.frontIndex]),
+		) as ISequencedDocumentMessage;
+
+		// TODO: Determine if this needs to be adapted for handling server-generated messages (which have null clientId and referenceSequenceNumber of -1).
+		// eslint-disable-next-line @typescript-eslint/no-unnecessary-type-assertion
+		this.minSeq.set(message.clientId as string, message.referenceSequenceNumber);
+		if (
+			this.runtimeOptions.flushMode === FlushMode.Immediate ||
+			this.lastProcessedMessage?.clientId !== message.clientId
+		) {
+			this.sequenceNumber++;
```

Comment:
this is incorrect. all client must get the same sequence number of all ops, i don't think this will do that, my suggestion will make this easier to manage